### DISCUSSION TITLE
BLE Retrieve device info & Retrieve settings

### DISCUSSION THREAD

**Discussion by jaqkar** [syssi/esphome-jk-bms#440] [Q&A] [answered]
I see Retrieve device info & Retrieve settings with PRESS next to them over BLE connection. Unclear what these two are doing or where to see the info. Any insights will be appreciated.

**Comment by syssi** [comment, marked as answer]
These are button entities. If you send the payload `PRESS`  to the command topic the `device info frame` or `settings frame` gets requsted. The decoder of the `device info frame` just prints some messages into the log:

```
  ESP_LOGI(TAG, "  Vendor ID: %s", std::string(data.begin() + 6, data.begin() + 6 + 16).c_str());
  ESP_LOGI(TAG, "  Hardware version: %s", std::string(data.begin() + 22, data.begin() + 22 + 8).c_str());
  ESP_LOGI(TAG, "  Software version: %s", std::string(data.begin() + 30, data.begin() + 30 + 8).c_str());
  ESP_LOGI(TAG, "  Uptime: %d s", jk_get_32bit(38));
  ESP_LOGI(TAG, "  Power on count: %d", jk_get_32bit(42));
  ESP_LOGI(TAG, "  Device name: %s", std::string(data.begin() + 46, data.begin() + 46 + 16).c_str());
  ESP_LOGI(TAG, "  Device passcode: %s", std::string(data.begin() + 62, data.begin() + 62 + 16).c_str());
  ESP_LOGI(TAG, "  Manufacturing date: %s", std::string(data.begin() + 78, data.begin() + 78 + 8).c_str());
  ESP_LOGI(TAG, "  Serial number: %s", std::string(data.begin() + 86, data.begin() + 86 + 11).c_str());
  ESP_LOGI(TAG, "  Passcode: %s", std::string(data.begin() + 97, data.begin() + 97 + 5).c_str());
  ESP_LOGI(TAG, "  User data: %s", std::string(data.begin() + 102, data.begin() + 102 + 16).c_str());
  ESP_LOGI(TAG, "  Setup passcode: %s", std::string(data.begin() + 118, data.begin() + 118 + 16).c_str());
```

If you retrieve the `settings frame` all setting entities (in general  `number` entities) gets updated. I haven't implemented a periodic update interval here because the BMS beeps on every received command. If you use the Android app the settings frame is requested (+ the BMS beeps once) as soon as you go to the settings view.

At the moment, if you change a config setting using one of the `number` entities you should retrieve the settings frame once manually to make sure all settings entities represent the current config/state of the BMS.
